sdk/lib/queue/queue.cc

```diff
@@ -0,0 +1,435 @@
+{
+	size_t bufferSize;
+	size_t allocSize;
+	bool   overflow =
+	  __builtin_mul_overflow(elementCount, elementSize, &bufferSize);
+	// We must be able to accurately represent the buffer, so round it up to a
+	// representable length.
+	bufferSize = CHERI::representable_length(bufferSize);
+	static constexpr size_t CounterSize = sizeof(uint32_t);
+	// Round up the size to be correctly aligned for the counters at the end
+	// (if necessary) and add two counters worth of space.
+	overflow |= __builtin_add_overflow(
+	  bufferSize,
+	  (2 * CounterSize) + (CounterSize - (bufferSize & (CounterSize - 1))),
+	  &allocSize);
+	if (overflow)
+	{
+		return -EINVAL;
+	}
+	// We need the counters to be able to run to double the queue size without
+	// hitting the high bits.  Error if this is the case.
+	//
+	// This should never be reached: a queue needs to be at least 512 MiB
+	// (assuming one-byte elements) to hit this limit.
+	if (((elementCount | (elementCount * 2)) &
+	     HighBitFlagLock::reserved_bits()) != 0)
+	{
+		return -EINVAL;
+	}
+
+	// Allocate the space for the queue.
+	Capability buffer{heap_allocate(timeout, heapCapability, allocSize)};
+	if (!buffer.is_valid())
+	{
+		return -ENOMEM;
+	}
+
+	Capability<std::atomic<uint32_t>> producer{
+	  buffer.cast<std::atomic<uint32_t>>()};
+	Capability<std::atomic<uint32_t>> consumer{
+	  buffer.cast<std::atomic<uint32_t>>()};
+	// Make the producer and consumer point after the buffer
+	producer.address() += bufferSize;
+	consumer.address() += bufferSize + CounterSize;
+	// Set their bounds to 4 bytes.
+	producer.bounds() = CounterSize;
+	consumer.bounds() = CounterSize;
+	// The pointer used to free the allocation
+	*outAllocation  = buffer;
+	buffer.bounds() = bufferSize;
+	Debug::log("Created queue with buffer: {}", buffer);
+	// The handle
+	*outQueue = {elementSize, elementCount, buffer, producer, consumer};
+
+	return 0;
+}
+
+int queue_send(Timeout *timeout, struct QueueHandle *handle, const void *src)
+{
+	Debug::log("Send called on: {}", handle);
+	auto *producer   = handle->producer;
+	auto *consumer   = handle->consumer;
+	bool  shouldWake = false;
+	{
+		Debug::log("Lock word: {}", producer->load());
+		HighBitFlagLock l{*producer};
+		if (LockGuard g{l, timeout})
+		{
+			uint32_t producerCounter = counter_load(producer);
+			uint32_t consumerCounter = counter_load(consumer);
+			Debug::log("Producer counter: {}, consumer counter: {}, Size: {}",
+			           producerCounter,
+			           consumerCounter,
+			           handle->queueSize);
+			while (is_full(handle->queueSize, producerCounter, consumerCounter))
+			{
+				if (consumer->wait(timeout, consumerCounter) == -ETIMEDOUT)
+				{
+					Debug::log("Timed out on futex");
+					return -ETIMEDOUT;
+				}
+				consumerCounter = counter_load(consumer);
+			}
+			auto entry = buffer_at_counter(*handle, producerCounter);
+			if (int claim = heap_claim_fast(timeout, handle->buffer, src);
```

Comment:
Is this claim held indefinitely, until the next claim or compartment call? Same for the one in `queue_receive`.